### PR TITLE
[SPARK-47569][SQL] Disallow comparing variant.

### PR DESCRIPTION
**Pull request by chenhao-db**
### What changes were proposed in this pull request?

It adds type-checking rules to disallow comparing variant values (including group by a variant column). We may support comparing variant values in the future, but since we don't have a proper comparison implementation at this point, they should be disallowed on the user surface.

### How was this patch tested?

Unit tests.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/VariantSuite.scala [hunk 1]
```diff
@@ -192,4 +192,22 @@ class VariantSuite extends QueryTest with SharedSparkSession {
       }
     }
   }
+
+  test("group/order/join variant are disabled") {
```
Comment (gene-db) [inline sql/core/src/test/scala/org/apache/spark/sql/VariantSuite.scala:196]: Thanks for the tests!

What about sort by? or window partition by, or window order by? Some examples:

```
select parse_json('') sort by 1

with t as (select 1 as a, parse_json('') as v)
select rank() over (partition by v order by a)

with t as (select 1 as a, parse_json('') as v)
select rank() over (partition by a order by v)
```

**Review by gene-db — COMMENTED** [review]
@chenhao-db Thanks! I left a question.

File: sql/core/src/test/scala/org/apache/spark/sql/VariantSuite.scala (hunk 1 above)
Comment (chenhao-db) [reply sql/core/src/test/scala/org/apache/spark/sql/VariantSuite.scala:196]: This is a very good point. Actually, the third statement will not fail as expected, and it is really an error in the type check for `Window`. Both `partitionSpec` and `orderSpec` must be orderable, but the current type check only includes `orderSpec`. If the `partitionSpec` contains a map type, the query will fail later in an optimizer rule with `INTERNAL_ERROR`. I will create another PR to fix the type check for `Window`.

Comment (chenhao-db) [reply sql/core/src/test/scala/org/apache/spark/sql/VariantSuite.scala:196]: If we can merge https://github.com/apache/spark/pull/45730 before this PR, we can add the second statement to the tests.

**Review by gene-db — APPROVED** [review]
@chenhao-db Thanks for fixing and the test cases!

LGTM

**Comment by chenhao-db** [comment]
@cloud-fan Could you help review this PR? Thanks!

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**